src/plugins/timeframe-selector-plugin.ts

```diff
@@ -265,133 +289,240 @@ export class TimeframeSelectorPlugin extends UIPlugin<TimeframeSelectorPluginSty
-                    sort: 0,
-                    level: level,
-                    index,
-                    type: 'start',
-                },
-                {
-                    pos: start,
-                    sort: 1,
-                    level: level + 1,
-                    index,
-                    type: 'start',
-                },
-                {
-                    pos: end,
-                    sort: 2,
-                    level: level + 1,
-                    index,
-                    type: 'end',
-                },
-                {
-                    pos: end,
-                    sort: 3,
+            if (type === 'end') {
+                renderDots.push({
+                    pos: time,
                     level: level,
-                    index,
-                    type: 'end',
-                }
-            );
-        });
-
-        this.dots = dots.sort((a, b) => {
-            if (a.pos !== b.pos) {
-                return a.pos - b.pos;
+                });
             }
 
-            if (a.index === b.index) {
-                return a.sort - b.sort;
+            if (type === 'start') {
+                level++;
+            } else {
+                level--;
             }
 
-            if (a.type === 'start' && b.type === 'start') {
-                return a.level - b.level;
-            } else if (a.type === 'end' && b.type === 'end') {
-                return b.level - a.level;
-            }
+            maxLevel = Math.max(maxLevel, level);
 
-            return 0;
+            renderDots.push({
+                pos: time,
+                level,
+            });
         });
 
-        this.maxLevel = maxLevel;
+        return {
+            dots: renderDots,
+            maxLevel,
+        };
+    }
+
+    makeWaterfallDots() {
+        if (this.waterfall) {
+            const data = parseWaterfall(this.waterfall);
+
+            const intervals = Object.entries(
+                data.reduce((acc: Record<string, PreparedWaterfallInterval[]>, { intervals }) => {
+                    intervals.forEach((interval) => {
+                        if (!acc[interval.color]) {
+                            acc[interval.color] = [];
+                        }
+
+                        acc[interval.color].push(interval);
+                    });
+
+                    return acc;
+                }, {})
+            );
+
+            const points = intervals.map(([color, intervals]) => {
+                const newPoints: { type: 'start' | 'end'; time: number }[] = [];
+
+                intervals.forEach(({ start, end }) => {
+                    newPoints.push({ type: 'start', time: start });
+                    newPoints.push({ type: 'end', time: end });
+                });
+
+                newPoints.sort((a, b) => a.time - b.time);
+
+                return {
+                    color,
+                    points: newPoints,
+                };
+            }, {});
```

Comment:
I am confused a bit here with the `, {}`
it is a map function on the array
`{}` is a `thisArg` parameter, but seems redundant?

Reply:
Probably forgot to delete. Just at the beginning I wrote a `reduce` here.